### ISSUE TITLE
SIGSEV When extensions is enabled with SourceGuardian

### ISSUE THREAD

**Issue by wsn-dev** [open-telemetry/opentelemetry-php-instrumentation#106]
**Describe your environment** 
Container built via a Dockerfile
`FROM php:8.0.30-alpine

ADD https://github.com/mlocati/docker-php-extension-installer/releases/latest/download/install-php-extensions /usr/local/bin/

RUN apk add --no-cache strace && chmod +x /usr/local/bin/install-php-extensions && \
    install-php-extensions sourceguardian  \
    opentelemetry-php/ext-opentelemetry@1.0.0 \
    grpc \
    mbstring \
    zlib \
    protobuf

ADD test.php /app/test.php
`

**Steps to reproduce**
1. Encode a test.php(content as simple as possible e.g. <?php echo 'test';) file via sourceguardian (https://www.sourceguardian.com/online-php-encoder.html choose PHP8+)
2. Place the encoded php content as test.php next to the Dockerfile 
3. Build the Dockerfile with: docker build -f Dockerfile -t otel-debug-local . 
4. Run the container with docker run --rm -it --entrypoint /bin/sh otel-debug-local
5. Call the php file with php /app/test.php

**What is the expected behavior?**
An output which says test

**What is the actual behavior?**
An error which says Segmentation fault

**Additional context**
The SourceGuardian extension does not come as a debug variant so i could not use gdb or a php debug variant.
Currently i only could provide an strace run see attached strace.txt
[strace.txt](https://github.com/open-telemetry/opentelemetry-php-instrumentation/files/13191031/strace.txt)

**Comment by wsn-dev** [comment]
Managed to get a minimal gdb 
0  0x00007f5681a02028 in ?? ()

1  0x000055ef3da4f37d in _zend_observe_fcall_begin (execute_data=0x7f5681a13dd0) at /usr/src/php-8.0.30/Zend/zend_observer.c:163

2  0x000055ef3da4f3fe in zend_observer_fcall_begin (execute_data=0x7f5681a13dd0) at /usr/src/php-8.0.30/Zend/zend_observer.c:176

3  0x000055ef3d9f02a5 in zend_execute (op_array=0x7f5681a8b100, return_value=0x7f5681a13d60) at /usr/src/php-8.0.30/Zend/zend_vm_execute.h:59522

4  0x00007f5681cab243 in ?? () from /usr/local/lib/php/extensions/no-debug-non-zts-20200930/sourceguardian.so

5  0x000055ef3d9d68e7 in ZEND_DO_ICALL_SPEC_RETVAL_USED_HANDLER () at /usr/src/php-8.0.30/Zend/zend_vm_execute.h:1295

6  execute_ex (ex=0x7f5681a13020) at /usr/src/php-8.0.30/Zend/zend_vm_execute.h:55163

7  0x000055ef3d9f02b5 in zend_execute (op_array=0x7f5681a8b000, return_value=0x0) at /usr/src/php-8.0.30/Zend/zend_vm_execute.h:59523

8  0x000055ef3d89b3f8 in zend_execute_scripts (type=8, retval=0x0, file_count=3) at /usr/src/php-8.0.30/Zend/zend.c:1694

9  0x000055ef3d7e0c0c in php_execute_script (primary_file=0x7fffaebdadf0) at /usr/src/php-8.0.30/main/main.c:2546

10 0x000055ef3da50efd in do_cli (argc=6, argv=0x55ef3f1061a0) at /usr/src/php-8.0.30/sapi/cli/php_cli.c:949

11 0x000055ef3da520a9 in main (argc=6, argv=0x55ef3f1061a0) at /usr/src/php-8.0.30/sapi/cli/php_cli.c:1337

**Comment by brettmc** [comment]
> What is your opinion should i create a issue @ the php-src repo ?

I wouldn't do that just yet. zend_observer calls opentelemetry functions which are registered on startup, so it's very likely that it's code in the opentelemetry extension which is causing the segfault.
It's quite annoying that you can't run their extension against a debug build of PHP - I was able to replicate your issue, but not get any useful traces. Perhaps if you were to reach out to the sourceguardian vendor they might be able to help run it with debug symbols?

**Comment by wsn-dev** [comment]
Thank you @brettmc i reached out to sourceguardian about it. 
I try to keep this thread in sync once i receive a feeback. I also got a different segfault with a simple demo extension for zend_observer calls (https://github.com/SammyK/observer) 
`0x000055a8e93d914b in zend_vm_calc_used_stack (func=0x2, num_args=0) at /usr/src/php-8.0.30/Zend/zend_execute.h:224
224             if (EXPECTED(ZEND_USER_CODE(func->type))) {`

**Comment by brettmc** [comment]
> I try to keep this thread in sync once i receive a feeback. I also got a different segfault with a simple demo extension for zend_observer calls (https://github.com/SammyK/observer)

Good thinking! It's a pretty simple PoC, so maybe that points to SourceGuardian not being compatible with the observer API? It's pretty new to PHP and probably not very widely used yet.

**Comment by wsn-dev** [comment]
@brettmc This is also what i think. 
At the moment i only got the feedback that SourceGuardian encoded files should be used on production system(Which indeed does not help in debuging it and I am still in contact with SourceGuardian).
What i reached is the following at the moment:

I stripped down the otel_observer.c function so that the begin ,end and init function are only executing the following:

static zend_observer_fcall_handlers
observer_fcall_init(zend_execute_data *execute_data) {
    zend_observer_fcall_handlers handlers = {NULL, NULL};
    handlers.begin = observer_begin_handler;
    return handlers;
}

static void observer_begin_handler(zend_execute_data *execute_data) {
    php_printf("[BEGIN %s()]\n","observer_begin_handler");
}

static void observer_end_handler(zend_execute_data *execute_data,
                                 zval *retval) {
    php_printf("[END %s()]\n","observer_end_handler");
}

Same as the POC but with the difference that i get an output before the segfault:
[BEGIN observer_begin_handler()] which points me to a php-src thing.

But still the same segfault. 
1 0x000055ef3da4f37d in _zend_observe_fcall_begin (execute_data=0x7f5681a13dd0) at /usr/src/php-8.0.30/Zend/zend_observer.c:163
2 0x000055ef3da4f3fe in zend_observer_fcall_begin (execute_data=0x7f5681a13dd0) at /usr/src/php-8.0.30/Zend/zend_observer.c:176

**Comment by wsn-dev** [comment]
@brettmc 
I have received the following response from the support team at Sourceguardian.

> I doubt this will work because using of any tracing tools and other extensions which deeply integrate with PHP and affect how bytecode is processed, will not work since this affects SourceGuardian from loading and executing the encoded bytecode.

So i think there should be some type of remark in this project like it was for blackfire #80.

Another interesting thing to note is that there is a statement regarding the Sourceguardian Extension on the Blackfire [Support website](https://support.blackfire.platform.sh/hc/en-us/articles/4843000788882-Known-incompatibilities-with-the-Blackfire-PHP-probe), which states the following:

> PHP is likely to crash when the following PHP extensions are installed together with Blackfire:
XHProf
Pinba
Suhosin
Sourceguardian
Ioncube

**Comment by brettmc** [comment]
Creating a conflict with SourceGuardian is straight-forward, it turns out. I couldn't work out why I couldn't see the zend extensions from C code, but https://www.phpinternalsbook.com/php7/extensions_design/zend_extensions.html#zend-extensions-lifetime-hooks says that modules are all loaded before zend_extensions are (SourceGuardian is a zend_extension).

Since we do our checking in `MINIT()`, that's too early to be able to see zend extensions. I tried moving the checks into `RINIT()` but that breaks almost all of our tests (but the conflict checking worked, so that's something).

So, it looks like for conflicts with zend extensions, the best we can do is to document it.

Also, I tried running with xhprof and ionCube enabled (which were the easiest for me to install), and it didn't obviously break.
